packages/common/src/EthersTransactionUtils.ts

```diff
@@ -0,0 +1,241 @@
+ * - On `UNPREDICTABLE_GAS_LIMIT`, rethrows the same error object and annotates it with `error.revertData` (hex).
+ * - Optionally applies a gas limit multiplier (percentage), e.g. `110` => +10%.
+ * - Validates that `originalTx.to` matches `contract.address` and that `data` is present.
+ * - Re-encodes and **dispatches through the contract ABI** (`contract.functions[name](...args, overrides)`), so that
+ *   `tx.wait()` resolves to a **ContractReceipt with decoded `events`**.
+ * - The returned `tx.wait()` is overridden so that if the tx **mines and reverts** (ethers throws `CALL_EXCEPTION`),
+ *   the function replays the call at `receipt.blockNumber` to recover the revert payload and annotates the thrown
+ *   error with `error.revertData` before rethrowing the **same** error instance.
+ * Note: Enhanced error data would only work for contracts backed by JsonRpcProvider, as it relies on the specific
+ *   placement of the revert data in the thrown error structure. Other providers may not support this.
+ *
+ * @param contract             Ethers v5 Contract instance (must have a signer attached).
+ * @param originalTx           Populated transaction targeting `contract.address` with encoded `data`.
+ * @param gasLimitMultiplier   Optional percentage multiplier for the estimated gas limit (e.g., 110 = +10%).
+ * @returns                    A `ContractTransaction` whose `wait()` returns a `ContractReceipt` with decoded events,
+ *                             or throws `CALL_EXCEPTION` with `error.revertData` on failure.
+ */
+export async function runEthersContractTransaction(
+  contract: Contract,
+  originalTx: PopulatedTransaction,
+  gasLimitMultiplier?: number
+): Promise<ContractTransaction> {
+  const { signer } = contract;
+  if (!Signer.isSigner(signer)) throw new Error("Contract has invalid signer");
+
+  // Estimate the gas limit adding the revert data on failure.
+  const submittedTx = await estimateTxGasLimit(signer, originalTx, gasLimitMultiplier);
+
+  // Validate this can be run as a contract transaction.
+  if (submittedTx.to?.toLowerCase() !== contract.address.toLowerCase()) {
+    throw new Error(`Transaction 'to' address ${submittedTx.to} does not match contract address ${contract.address}.`);
+  }
+  if (submittedTx.data === undefined || submittedTx.data === "0x") {
+    throw new Error("Transaction 'data' is undefined or empty, cannot run contract transaction.");
+  }
+
+  // Extract transaction overrides and parse the transaction data.
+  const overrides = (({ to: _to, from: _from, data: _data, chainId: _chainId, ...overrides }) => overrides)(
+    submittedTx
+  );
+  const parsed = contract.interface.parseTransaction({ data: submittedTx.data });
+
+  // Submit the transaction re-encoding with contract interface and applying overrides.
+  const tx = (await contract.functions[parsed.name](...parsed.args, overrides)) as ContractTransaction;
```

Comment:
Should we use `parsed.signature` instead to enable handling function overloading?

Reply:
good catch, fixed